R/mod_hierarchical_count_table.R

```diff
@@ -23,9 +47,159 @@ EC <- poc( # nolint
+                         recurse = TRUE) {
+
+  # Flags for when dates do not exist on population and event data frames
+  has_origin_dt <- origin_date_var %in% names(pop_df)
+  has_censor_dt <- censor_date_var %in% names(pop_df)
+  has_event_dt <- event_date_var %in% names(event_df)
+
+  # Remove rows where hierarchy value is NA in any processed hierarchy columns
+  event_df <- event_df |>
+    dplyr::filter(dplyr::if_all(dplyr::all_of(hierarchy), ~ !is.na(.x)))
+
+  # For each hierarchy group, only keep first event occurrence
+  if (has_event_dt && nrow(event_df) > 0) {
+    event_df2 <- event_df |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(c(subjid_var, hierarchy)))) |>
+      dplyr::summarise(!!event_date_var := min(.data[[event_date_var]]), .groups = "drop")
+  } else {
+    event_df2 <- event_df |>
+      dplyr::group_by(dplyr::across(dplyr::all_of(c(subjid_var, hierarchy)))) |>
+      dplyr::slice(1) |>
+      dplyr::ungroup()
+  }
+
+  # Identify records with events before merging onto all hierarchy combinations
+  event_df2[["CNSR"]] <- 0
+
+  # Expand population data using grid of hierarchy combinations to create base records for ADTTE ----
+
+  if (length(hierarchy) != 0) {
+    hierarchy_grid <- unique(event_df2[, hierarchy, drop = FALSE])
+    adtte <- dplyr::cross_join(pop_df, hierarchy_grid)
+  } else {
+    adtte <- pop_df
+  }
+
+  # Merge event dates onto base records of ADTTE ----
+
+  adtte <- dplyr::left_join(adtte, event_df2, by = c(subjid_var, hierarchy))
+
+  # Derive ADTTE variables ----
+
+  hierarchy_lvl <- length(hierarchy)
+
+  if (recurse) {
+    # Prepend hierarchy level to parameter value
+    adtte <- adtte |>
+      dplyr::mutate(PARAM = paste0("L", hierarchy_lvl, ": ", param_val),
+                    PARAMCD = paste0("L", hierarchy_lvl, "_", paramcd_val))
+  } else {
+    adtte <- adtte |>
+      dplyr::mutate(PARAM = param_val,
+                    PARAMCD = paramcd_val)
+  }
+
+  if (has_origin_dt) adtte[["STARTDT"]] <- adtte[[origin_date_var]]
+
+  if (has_event_dt && has_censor_dt) {
+    adtte[["ADT"]] <- pmin(adtte[[event_date_var]], adtte[[censor_date_var]], na.rm = TRUE)
+    if (has_origin_dt) adtte[["AVAL"]] <- as.numeric(adtte[["ADT"]] - adtte[["STARTDT"]] + 1)
+  }
+
+  adtte[["CNSR"]][is.na(adtte[["CNSR"]])] <- 1
+
+  adtte <- adtte |>
+    dplyr::select(dplyr::any_of(c(subjid_var, hierarchy, group_var,
+                                  "PARAM", "PARAMCD", "AVAL", "STARTDT", "ADT", "CNSR")))
+
+  # Add hierarchy level to data
+  hier_lvl_col <- paste0(EC$VAL$SPECIAL_CHAR, "lvl")
+  adtte <- adtte |>
+    dplyr::mutate(!!hier_lvl_col := hierarchy_lvl)
+
+  # Recursive case - call function again with one less hierarchy level ----
+
+  if (recurse && hierarchy_lvl > 0) {
+    recurse_adtte <- create_adtte(event_df = event_df,
+                                  pop_df = pop_df,
+                                  hierarchy = hierarchy[-hierarchy_lvl],
```

Comment:
By looking at the structure, maybe it would be possible to write the iterative version instead of the recursive one?